balancer-js/src/modules/liquidity-managment/migrations/helpers.ts

```diff
@@ -0,0 +1,167 @@
+    {
+      poolId: from.id,
+      assetIn: from.address,
+      assetOut: mainToken.address,
+    },
+    {
+      poolId: to.id,
+      assetIn: mainToken.address,
+      assetOut: to.address,
+    },
+  ];
+
+  return path;
+};
+
+/**
+ * Converts Subgraph Pool to MigrationPool
+ * Recursively builds tokens
+ *
+ * @param id
+ * @param repository
+ * @returns
+ */
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export const buildMigrationPool = async (
+  id: string,
+  repository: Findable<Pool, PoolAttribute>
+) => {
+  const pool = await repository.find(id);
+  if (!pool) throw `Pool ${id} not found`;
+
+  const findTokens = async (token: string, parentPool: string) => {
+    let tokens: Array<MigrationPool> = [{ address: token }];
+    const pool = await repository.findBy('address', token);
+    if (pool && token != parentPool) {
+      const sortedTokens = pool.tokens.sort(cmpTokens);
+      const nestedTokens = await Promise.all(
+        sortedTokens.map(({ address }) => findTokens(address, pool.address))
+      );
+      tokens = [
+        {
+          address: token,
+          id: pool.id,
+          poolType: pool.poolType,
+          poolTypeVersion: pool.poolTypeVersion,
+          mainIndex: pool.mainIndex,
+          tokens: nestedTokens.flat(),
+        },
+      ];
+    }
+    return tokens;
+  };
+
+  // Sorts amounts out into ascending order (referenced to token addresses) to match the format expected by the Vault.
+  const sortedTokens = pool.tokens.sort(cmpTokens);
+
+  return {
+    id,
+    address: pool.address,
+    tokens: (
+      await Promise.all(
+        sortedTokens.map(({ address }) => findTokens(address, pool.address))
+      )
+    ).flat(),
+    poolType: pool.poolType,
+    poolTypeVersion: pool.poolTypeVersion,
+    mainIndex: pool.mainIndex,
+  };
+};
+
+const cmpTokens = (tokenA: MigrationPool, tokenB: MigrationPool): number =>
+  tokenA.address.toLowerCase() > tokenB.address.toLowerCase() ? 1 : -1;
+
+/**
+ * Decodes the relayer return value to get the minimum BPT out.
+ *
+ * @param relayerReturnValue
+ * @returns
+ */
+export const getMinBptOut = (relayerReturnValue: string): string => {
+  // Get last two positions of the return value, bptOut is the last one or the second to last one in case there is a gauge deposit
+  // join and gauge deposit are always 0x, so any other value means that's the bptOut
+  const multicallResult = balancerRelayerInterface.decodeFunctionResult(
+    'multicall',
+    relayerReturnValue
+  );
+
+  const minBptOut = multicallResult[0]
+    .slice(-2)
+    .filter((v: string) => v !== '0x');
+
+  // NOTICE: When swapping from Linear Pools, the swap will query for the current wrapped token rate.
+  // It is possible that the rate changes between the static call checking for the BPT out and the actual swap, causing it to fail.
+  // To avoid this, we can add a small buffer to the min BPT out amount. eg. 0.0000001% of the BPT amount.
```

Comment:
Maybe this should be a `slippage` type input param? There's a chance swaps, etc fail as normal so a slippage would prevent failed txs.

Reply:
Good take. I was thinking about it as well, but came to a conclusion that it's better to leave slippage out of this function and up to implementation, so it can be applied externally on top of the expected amount.